### ISSUE TITLE
Idefics2 raises shape error in the backward pass with gradient_checkpointing

### ISSUE THREAD

**Issue by chenzizhao** [huggingface/transformers#30301]
I get a shape error if I PEFT the latest Idefics2 model at head with gradient checkpointing. The symptom is identical to [this](https://stackoverflow.com/questions/77656467/attention-mask-error-when-fine-tuning-mistral-7b-using-transformers-trainer). Seems like the following block isn't compatible with gradient checkpointing.
https://github.com/huggingface/transformers/blob/e4ea19b958c89d61e42461fac6ac8441787121f8/src/transformers/models/idefics2/modeling_idefics2.py#L1593-L1596

In my local workspace a fix similar to https://github.com/huggingface/transformers/pull/28061/commits/91f4b7e2b8c305dfa9044463833ecbb0ae6ba560 removed the shape error but a quick review of the logic from a maintainer would be lovely. Happy to submit PR after. Thanks!

**Comment by chenzizhao** [comment]
Patch: https://github.com/huggingface/transformers/commit/a317323414387a25b8fcf4398494c4cf5f12afdc

**Comment by amyeroberts** [comment]
Hi @chenzizhao, thanks for raising this and providing links to the patch and relevant commits! 

Would you like to open a PR with these changes? This way you get the github contribution for the solution. 

cc @younesbelkada @gante for reference

**Comment by younesbelkada** [comment]
Nice catch @chenzizhao ! 
This is not catched by our CI sadly, we need to work on that ! adding it on my TODOs !
